### PR TITLE
Rework benchmark results aggregate postprocessing

### LINKED ISSUE #2306: It appears `fio-postprocess` is operating O(N2) or O(N3) processing log files
Place-holder issue based on evidence seen that for a 24-hr fio job, the processing of the resulting log files is O(N2) or O(N3), with an observed 1.4GB of resident memory for the process.

### PR DESCRIPTION
**Pull request by webbnh**
A large fio dataset brought to light significant performance problems in the benchmark results post-processing scripts.  The algorithms they use contain a number of O(n^2) behaviors which, when the dataset becomes large, operate with punishing slowness.  This pull request replaces the ordered list (array) used to contain time-series data for aggregation with a hash-table, indexed by timestamp.  This allows individual data to be accessed in constant or O(ln(n)) time, which allows the problematic dataset to be processed in minutes instead of...days?

The central changes are in `BenchProcessing.pm`.  However, the dataset structures are created in a custom manner by the individual benchmark post-processing scripts.  As such, in addition to `fio-postprocess`, this PR includes changes to `uperf-postprocess`, `dbench-postprocess`, and `trafficgen-postprocess`, as well, since they all use common functions.  In addition, `linpack-postprocess` got caught up in the crossfire, since it references `BenchProcessing.pm` although it doesn't use any of the affected functions.

This PR is a series of commits:
- The first is a point-fix to `fio-postprocess` to avoid re-writing a file during re-post-processing which should only be written after the original run.
- The next is another tweak to `fio-postprocess`, which allows to fail gracefully when sample-sets are missing from the data.
- The next is an addition to the unit tests for fio, which exercises an error case.
- The next commit is a straightforward reformatting of `process-iteration-samples` and `trafficgen-postprocess` -- this commit is intended only to improve the readability of these modules, and it contains minimal functional changes.
- The next commit is the meat of this PR.
- The next is whitespace changes to make the indentation of `uperf-postprocess` consistent.
- The last commit is trivial changes to `linpack-postprocess`.

_You may find it easier to review the changes one commit at a time, to separate the trivial (reformatting) changes from the substantive ones.  Alternatively, it may be helpful to have GitHub ignore whitespace changes._

These changes were hand-tested by re-processing specific benchmark results and by running the Agent unit tests.  For the most part, the reprocessed results are identical to the originals.  There was one value in the `fio-0` unit test which the old code appears to have calculated incorrectly; the new code produces the correct value, which in turn results in a change to the average value which includes it, and these values show up in three of the output files.

The other changes related to testing stem from imprecision in floating point numbers combined with a feature of Perl hashes.  By default, the ordering of values in hash tables in Perl is [intentionally non-deterministic](https://perldoc.perl.org/perlsec#Algorithmic-Complexity-Attacks).  Ordinarily this is not a problem, but because of partial-bit errors in floating point calculations, the non-determinism shows up as variations in the last digit of the floating point values, which causes the gold file comparisons to fail randomly.  To address this, the unit tests now disable the Perl hash seed randomization and bucket perturbation; however, doing so meant that several of the gold files needed to be updated, even though they were otherwise unaffected.

There is one caveat:  the scripts are written in a an adaptive way, such that if there are features missing from the dataset, the scripts simply skip over them when producing the output; thus, while I've tried to test my changes exhaustively, the results of that are only as good as the datasets which I was able to find.

This PR supersedes #2328.
Fixes #2306.

### CODE REVIEW

File: agent/bench-scripts/tests/pbench-fio/test-52.txt
Comment (portante) [inline agent/bench-scripts/tests/pbench-fio/test-52.txt:1]: Let's just punt this test entirely.

Ideally, this would be some functional test that is known to take a "long" time.

But we really want the unit tests to finish somewhat quickly.

Future work is going to move all this to the server side in Python3 anyways, so we'll want to solve such functional verification requirements other ways

File: agent/bench-scripts/pbench-fio
Comment (portante) [inline agent/bench-scripts/pbench-fio:869]: Some sort of tabs vs spaces thing happening here.

File: agent/bench-scripts/postprocess/uperf-postprocess
Comment (portante) [inline agent/bench-scripts/postprocess/uperf-postprocess:116]: We may have a tabs vs spaces thing happening here.

File: agent/bench-scripts/postprocess/process-iteration-samples
Comment (portante) [inline agent/bench-scripts/postprocess/process-iteration-samples:330]: May have a tabs vs spaces thing here.

File: agent/bench-scripts/postprocess/linpack-postprocess
Comment (portante) [inline agent/bench-scripts/postprocess/linpack-postprocess:48]: We may have a tabs vs spaces thing going on here.

File: agent/bench-scripts/postprocess/BenchPostprocess.pm
Comment (portante) [inline agent/bench-scripts/postprocess/BenchPostprocess.pm:22]: Should we track this in an issue?

**Review by portante — CHANGES_REQUESTED** [review]
Looking good.  I can't say that I have reviewed every code change in detail yet, but will dive in further when pulled out of `Draft` mode.

File: agent/bench-scripts/tests/pbench-fio/test-52.txt
Comment (webbnh) [reply agent/bench-scripts/tests/pbench-fio/test-52.txt:1]: With the changes in this PR, this test no longer takes very long to run, but it does nearly double the length of time required to run unit tests.  On that basis, plus the fact that it's not a "unit test", I'll pull it from this PR.

However, we should keep it and run it -- it's a valid and valuable functional test.  Where should I put it and how do I hook it up?

File: agent/bench-scripts/pbench-fio
Comment (webbnh) [reply agent/bench-scripts/pbench-fio:869]: What sort of thing?

As far as I can discern, the unchanged lines in this file are using hard-tabs for indentation, and the changed lines are doing the same.  (Using hard-tabs wouldn't be my preference, but my IDE handles them nicely, displaying a two-column indentation for them.)  I don't _think_ that I should be reformatting the file for this PR....

File: agent/bench-scripts/postprocess/uperf-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/uperf-postprocess:116]: Yep, that should be a tab, to match the surrounding code; will fix.

Comment (webbnh) [reply agent/bench-scripts/postprocess/uperf-postprocess:116]: Yup; fixed.  Thanks!

File: agent/bench-scripts/postprocess/process-iteration-samples
Comment (webbnh) [reply agent/bench-scripts/postprocess/process-iteration-samples:330]: Yup; fixed.

File: agent/bench-scripts/postprocess/linpack-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/linpack-postprocess:48]: This file was apparently using a mixed approach to indentation.  I seem to have converted it to use spaces rather than tabs.  Would you prefer that it used tabs like the other files?

File: agent/bench-scripts/postprocess/BenchPostprocess.pm
Comment (webbnh) [reply agent/bench-scripts/postprocess/BenchPostprocess.pm:22]: That's an interesting question.

In light of other conversations that we've had today, rather than draw resources and attention away from other endeavors, I'm inclined to ignore this mystery.

FWIW, `$script` is only referenced on the error paths of `get_cpubusy_series()`, which is somewhat less used than most functions in this module.  Moreover, I'm not up on Perl's scoping rules, but it's possible that the references to it will be satisfied by definition in the calling module rather than the containing one, in which case the lack of initialization here is unimportant.

If you like, I could revert this change....  🙂

File: agent/bench-scripts/pbench-fio
Comment (webbnh) [reply agent/bench-scripts/pbench-fio:869]: The "thing" happened a few lines later, and I fixed it.

**Comment by webbnh** [comment]
This PR is now ready for review.  (@ndokos and @dbutenhof, I don't know if GitHub sent you a notification....)

File: agent/bench-scripts/postprocess/trafficgen-postprocess
Comment (webbnh) [inline agent/bench-scripts/postprocess/trafficgen-postprocess:227]: This change is causing `test-22` to fail:  the previous code would omit any parameters with zero or empty-string values, whereas now they are included in the result.  This pulls in `"max_loss_pct"` and `"min_rate"`, which have zero values, which is probably appropriate (ditto `"trial_gap"`); however, it also pulls in `"pre_trial_cmd"`, which is an empty string, which might not be appropriate.  (The inclusion of `"max_loss_pct"` has the knock-on effect of causing the `"uid"` value to change.)

File: agent/bench-scripts/postprocess/trafficgen-postprocess
Comment (webbnh) [inline agent/bench-scripts/postprocess/trafficgen-postprocess:230]: This variable should be renamed to `$uid_value` or similar.  (To be consistent with other usage, a variable named `$uid_label` should contain the value returned from `get_label('uid_label')`.)

File: agent/bench-scripts/postprocess/trafficgen-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/trafficgen-postprocess:227]: Is this a bug or a feature?  (I.e., should I be reverting this change or updating the `test-22` gold file?)

File: agent/bench-scripts/tests/pbench-fio/test-52.txt
Comment (portante) [reply agent/bench-scripts/tests/pbench-fio/test-52.txt]: Is PR #2365 where this landed?

File: agent/bench-scripts/postprocess/BenchPostprocess.pm
Comment (portante) [reply agent/bench-scripts/postprocess/BenchPostprocess.pm:22]: No further changes needed for this.

File: agent/bench-scripts/postprocess/trafficgen-postprocess
Comment (portante) [reply agent/bench-scripts/postprocess/trafficgen-postprocess:227]: Sounds like a bug to me, but I am wondering if empty strings might be worth excluding.

File: agent/bench-scripts/tests/pbench-fio/test-52.txt
Comment (webbnh) [reply agent/bench-scripts/tests/pbench-fio/test-52.txt]: > Is PR #2365 where this landed?

Yup.

**Review by portante — COMMENTED** [review]
Interrim comments ...

File: agent/bench-scripts/postprocess/linpack-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/linpack-postprocess:48]: My apologies (I thought I did that, but...).

You can have GitHub turn off whitespace differences, which makes that module pretty straightforward to review.

File: agent/bench-scripts/postprocess/linpack-postprocess
Comment (ndokos) [inline agent/bench-scripts/postprocess/linpack-postprocess:85]: Is this debugging output? Should it be conditionally executed? Or commented out?

File: agent/bench-scripts/postprocess/uperf-postprocess
Comment (ndokos) [inline agent/bench-scripts/postprocess/uperf-postprocess:67]: Delete this line?

**Review by ndokos — COMMENTED** [review]
FIrst pass with a few comments. Still need to review the bulk of the changes: BenchPostprocess.pm, fio-postprocess, process-iterataion-samples, trafficgen-postprocess.

**Review by ndokos — DISMISSED** [review]
Minor nits here and there.

**Review by ndokos — COMMENTED** [review]
Test-22 (trafficgen) failed and it needs a rebase.

File: agent/bench-scripts/pbench-fio
Comment (webbnh) [reply agent/bench-scripts/pbench-fio:870]: I concur with the idea in general; however, in this case, in the next line we change the working directory to `$iteration_dir`, and in the line after that we reference `sample$sample.tar.xz`, so it makes reasonable sense to reference them both at line 870, rather than relying on the combination of them that we previously ensconced in `$benchmark_results_dir`.

File: agent/bench-scripts/pbench-fio
Comment (webbnh) [reply agent/bench-scripts/pbench-fio:894]: Yes.  (But, um, I'm afraid that you really don't get Perl -- this is Bash.  🤣)

File: agent/bench-scripts/postprocess/BenchPostprocess.pm
Comment (webbnh) [reply agent/bench-scripts/postprocess/BenchPostprocess.pm:376]: The two instances of the code appear to be consistent with each other (I presume they were cut-and-pasted from each other originally, and my changes to them didn't disrupt the original structure); however, the algorithm is very stateful, and the difference between the two instances is inside the loop.  I _could_ implement the abstraction with a callback function parameter which could be used to invoke the custom operation, but even that is a little tricky (and I'd have to figure out how to do it properly in Perl).

So, unless you insist, I'm going to decline.  (Someday, this will all be rewritten or removed, and we can address it then, hopefully with several other significant improvements.)

File: agent/bench-scripts/postprocess/fio-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/fio-postprocess:52]: The various `*-postprocess` scripts have at their cores a similar structure, including a series of top-level data structures.  However, not all of the scripts contain code to populate all of the data structures.  As such, I've commented out the declarations which are unused (for sound reasons...), but I've left them in place to maintain the commonality (and in case we should decide to enhance the code).

So, unless you insist, I'll decline.

File: agent/bench-scripts/postprocess/fio-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/fio-postprocess:64]: See [previous comment](https://github.com/distributed-system-analysis/pbench/pull/2349/files?file-filters%5B%5D=.csv&file-filters%5B%5D=.json&file-filters%5B%5D=.opts&file-filters%5B%5D=.sts&file-filters%5B%5D=.txt&file-filters%5B%5D=No+extension#r716930094).

File: agent/bench-scripts/postprocess/fio-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/fio-postprocess:254]: Good catch:  fixed (along with one other, mixed tab/space indentation).

File: agent/bench-scripts/postprocess/fio-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/fio-postprocess:337]: Fixed.

File: agent/bench-scripts/postprocess/uperf-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/uperf-postprocess:67]: See [subsequent comment](https://github.com/distributed-system-analysis/pbench/pull/2349#discussion_r716930094).

File: agent/bench-scripts/postprocess/linpack-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/linpack-postprocess:85]: I'm afraid that I have no idea; and, I don't see any obvious tests for this script, so I can't really comment.  But, I will say, since I didn't add this line, I'm not really inclined to remove it....

File: agent/bench-scripts/postprocess/trafficgen-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/trafficgen-postprocess:227]: So, @portante, this issue is now the sole blocker for this PR (at least, until the next round of code review...).  Should I

- revert this change,
- accept this change and update the gold file, or
- split the difference:  report zeros but omit empty strings?

File: agent/bench-scripts/postprocess/linpack-postprocess
Comment (portante) [reply agent/bench-scripts/postprocess/linpack-postprocess:48]: As a rule, while the tool helps to deal with whitespace changes, it is not something that we should allow.

Whitespace changes should be made in separate commits, and it is easy to do by inserting a commit that only makes whitespace changes via `git rebase -i` to edit them into the series of commits.

Comment (portante) [reply agent/bench-scripts/postprocess/linpack-postprocess:48]: Not every tool people use give you that same way of ignoring whitespace.

And in this case, GitHub gratuitously removes the "hide whitespace changes" when switching to conversations that reference whitespace changes.

**Review by portante — CHANGES_REQUESTED** [review]

File: agent/bench-scripts/postprocess/linpack-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/linpack-postprocess:48]: My apologies, again.  I did put the whitespace changes into separate commits for most of the modules which received significant amounts of whitespace change.  Somehow this module escaped that treatment.

Are you satisfied with the final state of this module?  (I.e., if so, could you mark this conversation "resolved"?)

File: agent/bench-scripts/pbench-fio
Comment (webbnh) [reply agent/bench-scripts/pbench-fio:870]: I don't find `$(dirname ${benchmark_result_dir})` to be easier to read or execute than `$iteration_dir`; likewise I don't think that `$(basename ${benchmark_result_dir}).tar.xz` is better than `sample${sample}.tar.xz`.

Under other circumstances, we might be inclined to cache the values returned by `dirname` and `basename` in variables, so, since we already have them, why not use them?

File: agent/bench-scripts/postprocess/trafficgen-postprocess
Comment (portante) [reply agent/bench-scripts/postprocess/trafficgen-postprocess]: Filed issue #2462 to track.

File: agent/bench-scripts/pbench-fio
Comment (portante) [reply agent/bench-scripts/pbench-fio:870]: Okay, that is fair given you are not making this any worse than it was, and hopefully we can drop the code entirely in the future.

File: agent/bench-scripts/postprocess/linpack-postprocess
Comment (portante) [reply agent/bench-scripts/postprocess/linpack-postprocess:48]: Okay, I'll be a softy and let this slide ... ;)

**Review by portante — DISMISSED** [review]

File: agent/bench-scripts/postprocess/fio-postprocess
Comment (ndokos) [inline agent/bench-scripts/postprocess/fio-postprocess:345]: "graped" ->"graphed"

**Review by ndokos — APPROVED** [review]
One typo in a comment, but otherwise LGTM. I believe I have resolved all previous conversations. Feel free to merge as is, or fix the typo and merge. I  will resolve the outstanding conversation in a little while, after you get a chance to see it.

File: agent/bench-scripts/postprocess/trafficgen-postprocess
Comment (webbnh) [reply agent/bench-scripts/postprocess/trafficgen-postprocess]: It turned out that reverting this change didn't resolve the problem.  🙁

So, I went ahead and added code to exclude fields with empty string values.  Unfortunately, there is considerably subtlety in Perl's type-less scalars, and the JSON formatter that we use disambiguates them based on how they were _last **used**_, so checking to see that a (numeric) value is a not an empty _string_ caused several of the numeric fields to then be dumped **as strings**.

So, I had to add logic to detect numeric values based on their contents, and, for things which looked numeric, to re-numerify them.  Unfortunately, JSON boolean values appear to be numeric, but we want them formatted as neither numbers nor strings, so I had to add further logic to ensure that they come through unscathed.

Nevertheless, after much wrangling, I am satisfied with the result.  It does add a few fields to the output, and the UID does change, but I think the latter is an intended result, and the former are not bad things.  (And, it aligns well with the `b0.69` results, although I'm not sure what we did in 0.71 which requires extra logic.)

**Comment by webbnh** [comment]
OK, I've made all the requested changes and squashed them into their respective commits (I'm intending to merge this branch via "Rebase and merge" and keep the remaining/existing commits separate).  So, once Mr. Jenkins blesses this, it should be good to go.

Could I please get two approvals, @ndokos, @portante, @dbutenhof?

**Review by ndokos — APPROVED** [review]

**Review by portante — APPROVED** [review]

**Pull request merged**
